### PR TITLE
✨ Implement ProcessStarted Notification

### PR DESCRIPTION
**Pull request by robpal94**
---

**Changes:**

1. Add a method which calls a callback when a ProcessInstance starts.

**Issues:**

Related: https://github.com/process-engine/process_engine_runtime/issues/134

PR: #26

---

See the [examples](#example) for inspiration.

## How can others test the changes?

> Describe how others can test your changes (you can remove this section for typo fixes etc.)

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Fixing Security Issues   | `:lock:`             | 🔒     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️      |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️      |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: src/accessors/external_accessor.ts [hunk 1]
```diff
@@ -61,7 +61,10 @@ export class ExternalAccessor implements IConsumerApiAccessor {
 
   public onProcessTerminated(identity: IIdentity, callback: Messages.CallbackTypes.OnProcessTerminatedCallback): void {
     this._ensureIsAuthorized(identity);
-    this._socket.on(socketSettings.paths.processTerminated, callback);
+  }
+
+  public onProcessStarted(identity: IIdentity, callback: Messages.CallbackTypes.OnProcessStartedCallback): void {
+    this._socket.on(socketSettings.paths.processStarted, callback);
```
[the remark this answers is not in the document]
Comment (SteffenKn) [reply src/accessors/external_accessor.ts]:
```suggestion
    this._ensureIsAuthorized(identity);
    this._socket.on(socketSettings.paths.processStarted, callback);
```

File: src/accessors/internal_accessor.ts [hunk 2]
```diff
@@ -47,6 +47,10 @@ export class InternalAccessor implements IConsumerApiAccessor {
     this._consumerApiService.onManualTaskFinished(identity, callback);
   }
 
+  public onProcessStarted(identity: IIdentity, callback: Messages.CallbackTypes.OnProcessStartedCallback): void {
+    this._consumerApiService.onProcessStarted(identity, callback);
```
[the remark this answers is not in the document]
Comment (SteffenKn) [reply src/accessors/internal_accessor.ts]:
```suggestion
    this._ensureIsAuthorized(identity);
    this._consumerApiService.onProcessStarted(identity, callback);
```

File: src/accessors/external_accessor.ts [hunk 3]
```diff
@@ -61,7 +61,10 @@ export class ExternalAccessor implements IConsumerApiAccessor {
 
   public onProcessTerminated(identity: IIdentity, callback: Messages.CallbackTypes.OnProcessTerminatedCallback): void {
     this._ensureIsAuthorized(identity);
```
Comment (SteffenKn) [inline src/accessors/external_accessor.ts]:
```suggestion
    this._ensureIsAuthorized(identity);
    this._socket.on(socketSettings.paths.processTerminated, callback);
```

File: src/accessors/external_accessor.ts [hunk 4]
```diff
@@ -61,7 +61,10 @@ export class ExternalAccessor implements IConsumerApiAccessor {
...
-    this._socket.on(socketSettings.paths.processTerminated, callback);
```
Comment (SteffenKn) [inline src/accessors/external_accessor.ts]: Was this intended? This looks broken now

File: src/accessors/external_accessor.ts [hunk 5]
```diff
@@ -64,6 +64,23 @@ export class ExternalAccessor implements IConsumerApiAccessor {
...
 
+  public onProcessStarted(identity: IIdentity, callback: Messages.CallbackTypes.OnProcessStartedCallback): void {
+    this._ensureIsAuthorized(identity);
+    this._socket.on(socketSettings.paths.processStarted, callback);
+  }
+
+  public onProcessWithProcessModelIdStarted(
+                                            identity: IIdentity,
+                                            callback: Messages.CallbackTypes.OnProcessStartedCallback,
+                                            processModelId: string): void {
```
Comment (ElRaptorus) [inline src/accessors/external_accessor.ts]: Please fix the indentation and alignment of these parameters.

File: src/accessors/internal_accessor.ts [hunk 6]
```diff
@@ -47,6 +47,19 @@ export class InternalAccessor implements IConsumerApiAccessor {
...
 
+  public onProcessStarted(identity: IIdentity, callback: Messages.CallbackTypes.OnProcessStartedCallback): void {
+    this._ensureIsAuthorized(identity);
+    this._consumerApiService.onProcessStarted(identity, callback);
+  }
+
+  public onProcessWithProcessModelIdStarted(
+                                            identity: IIdentity,
+                                            callback: Messages.CallbackTypes.OnProcessStartedCallback,
+                                            processModelId: string): void {
```
Comment (ElRaptorus) [inline src/accessors/internal_accessor.ts]: Like above, please fix indentation and alignment.

File: src/consumer_api_client_service.ts [hunk 7]
```diff
@@ -38,6 +38,17 @@ export class ConsumerApiClientService implements IConsumerApi {
...
   }
 
+  public onProcessStarted(identity: IIdentity, callback: Messages.CallbackTypes.OnProcessStartedCallback): void {
+    this.consumerApiAccessor.onProcessStarted(identity, callback);
+  }
+
+  public onProcessWithProcessModelIdStarted(
+                                            identity: IIdentity,
+                                            callback: Messages.CallbackTypes.OnProcessStartedCallback,
+                                            processModelId: string): void {
```
Comment (ElRaptorus) [inline src/consumer_api_client_service.ts]: See above.

File: src/accessors/external_accessor.ts [hunk 8]
```diff
@@ -64,6 +64,24 @@ export class ExternalAccessor implements IConsumerApiAccessor {
...
+  public onProcessStarted(identity: IIdentity, callback: Messages.CallbackTypes.OnProcessStartedCallback): void {
+    this._ensureIsAuthorized(identity);
+    this._socket.on(socketSettings.paths.processStarted, callback);
+  }
+
+  public onProcessWithProcessModelIdStarted(
+    identity: IIdentity,
+    callback: Messages.CallbackTypes.OnProcessStartedCallback,
+    processModelId: string,
+    ): void {
```
Comment (ElRaptorus) [inline src/accessors/external_accessor.ts]: Almost...

```suggestion
  ): void {
```

**Pull request merged**